src/adapter/src/coord/timestamp_selection.rs

```diff
@@ -507,36 +508,27 @@ pub trait TimestampProvider {
         &self,
         id_bundle: &CollectionIdBundle,
         compute_instance: ComputeInstanceId,
+        read_holds: &ReadHolds<mz_repr::Timestamp>,
         candidate: mz_repr::Timestamp,
     ) -> String {
-        let invalid_indexes =
-            if let Some(compute_ids) = id_bundle.compute_ids.get(&compute_instance) {
-                compute_ids
-                    .iter()
-                    .filter_map(|id| {
-                        let since = self.compute_read_frontier(compute_instance, *id).to_owned();
-                        if since.less_equal(&candidate) {
-                            None
-                        } else {
-                            Some(since)
-                        }
-                    })
-                    .collect()
-            } else {
-                Vec::new()
-            };
-        let invalid_sources = id_bundle.storage_ids.iter().filter_map(|id| {
-            let since = self.storage_read_capabilities(*id).to_owned();
-            if since.less_equal(&candidate) {
-                None
-            } else {
-                Some(since)
+        let mut invalid = Vec::new();
+
+        if let Some(compute_ids) = id_bundle.compute_ids.get(&compute_instance) {
+            for id in compute_ids {
+                let since = read_holds.since(id);
+                if !since.less_equal(&candidate) {
+                    invalid.push((*id, since));
+                }
             }
-        });
-        let invalid = invalid_indexes
-            .into_iter()
-            .chain(invalid_sources)
-            .collect::<Vec<_>>();
+        }
+
+        for id in id_bundle.storage_ids.iter() {
+            let since = read_holds.since(id);
+            if !since.less_equal(&candidate) {
+                invalid.push((*id, since));
+            }
+        }
+
         format!(
```

Comment:
The error is peculiar: before, it was including the Debug representation of `invalid`, which was a `Vec<Antichain<T>>`. I changed that to a `Vec<(GlobalId, Antichain<T>)>`, which seems equally bad, but at least provides more information about which IDs had the offending sinces. Thoughts?